sympy/core/mul.py

```diff
@@ -195,6 +195,10 @@ def flatten(cls, seq):
                                 bargs.insert(0, ar)
                             bargs = [Add._from_args(bargs)]
                             rv = bargs, [], None
+            if a.is_commutative and b.is_commutative:
+                if ((a.is_finite == False and b.is_zero) or
+                        (b.is_finite == False and a.is_zero)):
+                    rv = [S.NaN], [], None
```

Comment:
Worth experimenting with leaving it unevaluated rather than returning NaN? (just a thought, I've really only glanced at this).